src/client-side-encryption/autoEncrypter.ts

```diff
@@ -0,0 +1,641 @@
+    const encryption = getMongoDBClientEncryption();
+    if ('kModuleError' in encryption) {
+      throw encryption.kModuleError;
+    }
+    return encryption.MongoCrypt;
+  }
+
+  /**
+   * Create an AutoEncrypter
+   *
+   * **Note**: Do not instantiate this class directly. Rather, supply the relevant options to a MongoClient
+   *
+   * **Note**: Supplying `options.schemaMap` provides more security than relying on JSON Schemas obtained from the server.
+   * It protects against a malicious server advertising a false JSON Schema, which could trick the client into sending unencrypted data that should be encrypted.
+   * Schemas supplied in the schemaMap only apply to configuring automatic encryption for Client-Side Field Level Encryption.
+   * Other validation rules in the JSON schema will not be enforced by the driver and will result in an error.
+   *
+   * @example <caption>Create an AutoEncrypter that makes use of mongocryptd</caption>
+   * ```ts
+   * // Enabling autoEncryption via a MongoClient using mongocryptd
+   * const { MongoClient } = require('mongodb');
+   * const client = new MongoClient(URL, {
+   *   autoEncryption: {
+   *     kmsProviders: {
+   *       aws: {
+   *         accessKeyId: AWS_ACCESS_KEY,
+   *         secretAccessKey: AWS_SECRET_KEY
+   *       }
+   *     }
+   *   }
+   * });
+   * ```
+   *
+   * await client.connect();
+   * // From here on, the client will be encrypting / decrypting automatically
+   * @example <caption>Create an AutoEncrypter that makes use of libmongocrypt's CSFLE shared library</caption>
+   * ```ts
+   * // Enabling autoEncryption via a MongoClient using CSFLE shared library
+   * const { MongoClient } = require('mongodb');
+   * const client = new MongoClient(URL, {
+   *   autoEncryption: {
+   *     kmsProviders: {
+   *       aws: {}
+   *     },
+   *     extraOptions: {
+   *       cryptSharedLibPath: '/path/to/local/crypt/shared/lib',
+   *       cryptSharedLibRequired: true
+   *     }
+   *   }
+   * });
+   * ```
+   *
+   * await client.connect();
+   * // From here on, the client will be encrypting / decrypting automatically
+   */
+  constructor(client: MongoClient, options: AutoEncryptionOptions) {
+    this._client = client;
+    this._bypassEncryption = options.bypassAutoEncryption === true;
+
+    this._keyVaultNamespace = options.keyVaultNamespace || 'admin.datakeys';
+    this._keyVaultClient = options.keyVaultClient || client;
+    this._metaDataClient = options.metadataClient || client;
+    this._proxyOptions = options.proxyOptions || {};
+    this._tlsOptions = options.tlsOptions || {};
+    this._kmsProviders = options.kmsProviders || {};
+
+    const mongoCryptOptions: MongoCryptOptions = {};
+    if (options.schemaMap) {
+      mongoCryptOptions.schemaMap = Buffer.isBuffer(options.schemaMap)
+        ? options.schemaMap
+        : (serialize(options.schemaMap) as Buffer);
```

Comment:
You'll notice a lot of type casts here.  `schemaMap` explicitly needs to be a buffer for our bindings, but our bson serialize is platform-compatible and is typed to return `Uint8Arrays`.  At runtime, BSON returns `Buffer`s for Node, so it's still valid, but these casts are unavoidable.